Migrate BeamSpotDeviceProducer and HcalDigisSoAProducer to rely on implicit host-to-device copy

makortel opened the pull request:
#### PR description:

Making use of the functionality added in https://github.com/cms-sw/cmssw/pull/47111

Resolves https://github.com/cms-sw/framework-team/issues/1209

#### PR validation:

Code compiles

makortel committed:
Make BeamSpotDeviceProducer to rely on implicit host-to-device copy

makortel committed:
Make HcalDigisSoAProducer to rely on implicit host-to-device copy

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"bed107f110c74d2ff3b0daea3fc8bde9d376ce5e":{"files":["EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc"],"squashed":false,"time":1738331521},"fd1b754ccc820098dc31a14de73ae7b041096d7f":{"files":["RecoVertex/BeamSpotProducer/plugins/alpaka/BeamSpotDeviceProducer.cc"],"squashed":false,"time":1738330778}},"emoji":{},"last_seen_sha":"bed107f110c74d2ff3b0daea3fc8bde9d376ce5e","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47226/43506

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc  modified in PR(s): #47216
  - File  RecoVertex/BeamSpotProducer/plugins/alpaka/BeamSpotDeviceProducer.cc  modified in PR(s): #47216

cmsbuild commented:
A new Pull Request was created by  @makortel for master.

It involves the following packages:

- EventFilter/HcalRawToDigi (**reconstruction**)
- RecoVertex/BeamSpotProducer (**reconstruction, alca**)


@atpathak, @cmsbuild, @consuegs, @jfernan2, @mandrenguyen, @perrotta can you please review it and eventually sign? Thanks.
@GiacomoSguazzoni, @Martin-Grunewald, @VinInn, @VourMa, @abdoulline, @bsunanda, @dgulhan, @fabiocos, @francescobrivio, @martinamalberti, @missirol, @mmusich, @mtosi, @rovere, @rsreds, @tocheng, @yuanchao this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
enable gpu

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9750de/44108/summary.html
**COMMIT**: bed107f110c74d2ff3b0daea3fc8bde9d376ce5e
**CMSSW**: CMSSW_15_0_X_2025-01-31-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/47226/44108/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-31-1100+9750de/66693/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 36 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-31-1100+9750de/66693/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4130681
 - DQMHistoTests: [Total failures: 1348](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-31-1100+9750de/66693/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4129313
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 222 log files, 194 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-31-1100+9750de/66693/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-01-31-1100+9750de/66695/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53071
 - DQMHistoTests: [Total failures: 53](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-01-31-1100+9750de/66695/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 53018
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-01-31-1100+9750de/66695/triggerResults): no differences found

jfernan2 commented:
+1

makortel commented:
assign heterogeneous

cmsbuild commented:
New categories assigned: heterogeneous

@fwyzard,@makortel you have been requested to review this Pull request/Issue and eventually sign? Thanks

makortel commented:
CPU comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803 and https://github.com/cms-sw/cmssw/issues/47071

GPU comparison differences are in pixel tracking, except 12834.422 and 12834.423 show differences in ParticleFlow. @fwyzard Do we have a known non-reproducibility in PF GPU code?

fwyzard commented:
> Do we have a known non-reproducibility in PF GPU code?

Not that I remember.

@jsamudio ?

fwyzard commented:
@makortel the changes are fine.
Do you mind if I run some benchmarks (next week) before signing ?

makortel commented:
> Do you mind if I run some benchmarks (next week) before signing ?

No, go ahead. It would be nice to get this in pre3, but a backport PR would be trivial as well.

makortel commented:
> > Do we have a known non-reproducibility in PF GPU code?
> 
> Not that I remember.
> 
> @jsamudio ?

I opened an issue https://github.com/cms-sw/cmssw/issues/47233 as I saw these also in another test

perrotta commented:
+alca

makortel commented:
hold

https://github.com/cms-sw/cmssw/issues/47233#issuecomment-2629611455 showed this PR actually leads to a change (details still to be investigated)

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

makortel commented:
unhold

> [#47233 (comment)](https://github.com/cms-sw/cmssw/issues/47233#issuecomment-2629611455) showed this PR actually leads to a change (details still to be investigated)

The cause has been identified elsewhere (https://github.com/cms-sw/cmssw/issues/47233#issuecomment-2633541829)

(although now there is a merge conflict, effectively need to wait for the next IB)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47226/43590

cmsbuild commented:
Pull request #47226 was updated. @atpathak, @cmsbuild, @consuegs, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @perrotta can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_1_X. Please open a backport if it should also go in to CMSSW_15_0_X.

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9750de/44233/summary.html
**COMMIT**: b67acfa8321edfc18066889fc351e17ceab663d1
**CMSSW**: CMSSW_15_0_X_2025-02-05-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47226/44233/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-05-2300+9750de/66804/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 365 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-05-2300+9750de/66804/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4130882
 - DQMHistoTests: [Total failures: 6447](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-05-2300+9750de/66804/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4124415
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 222 log files, 194 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-02-05-2300+9750de/66804/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 24 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-05-2300+9750de/66801/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53071
 - DQMHistoTests: [Total failures: 862](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-05-2300+9750de/66801/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 52209
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_0_X_2025-02-05-2300+9750de/66801/triggerResults): no differences found

makortel commented:
CPU comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803 and https://github.com/cms-sw/cmssw/issues/47071

GPU comparison differences look compatible with non-reproducibilities in pixel code

perrotta commented:
@makortel I can re-sign after the rebase. However I notice that you could also make some further cleanup, if you think so:
- [L1](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L1) and [L2](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L2) in EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc are duplicated
- The header in [L4](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L4) can be now removed

makortel commented:
Thanks @perrotta!

> * [L1](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L1) and [L2](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L2) in EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc are duplicated

I removed one of these

> * The header in [L4](https://github.com/cms-sw/cmssw/blob/b67acfa8321edfc18066889fc351e17ceab663d1/EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisSoAProducer.cc#L4) can be now removed

Actually no, that header needs to be kept. The implicit host-to-device copy is still done by the `HcalDigisSoAProducer`, but only if the base class code sees the specialization of `cms::alpakatools::CopyToDevice<hcal::Phase[01DigiHostCollection>`, that is brought by that `#include`.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47226/43616

cmsbuild commented:
Pull request #47226 was updated. @atpathak, @cmsbuild, @consuegs, @fwyzard, @jfernan2, @makortel, @mandrenguyen, @perrotta can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9750de/44272/summary.html
**COMMIT**: b719a9aca9086491085d5fbb9bac7993f7daf908
**CMSSW**: CMSSW_15_1_X_2025-02-07-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47226/44272/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-07-1100+9750de/66832/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-07-1100+9750de/66832/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4130882
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-07-1100+9750de/66832/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4130862
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 222 log files, 194 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-07-1100+9750de/66832/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 24 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_1_X_2025-02-07-1100+9750de/66838/validateJR.html)
 - DQMHistoTests: Total files compared: 7
 - DQMHistoTests: Total histograms compared: 53071
 - DQMHistoTests: [Total failures: 869](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_1_X_2025-02-07-1100+9750de/66838/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 52202
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 6 files compared)
 - Checked 24 log files, 30 edm output root files, 7 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_15_1_X_2025-02-07-1100+9750de/66838/triggerResults): no differences found

perrotta commented:
+alca
- Signing it again after the rebase

jfernan2 commented:
+1

makortel commented:
@fwyzard Have you had a chance to run the tests?

makortel commented:
@fwyzard ping

makortel commented:
@fwyzard ping

fwyzard commented:
OK, these changes do not have any measurable impact on the HLT, neither on the full menu nor on the alpaka-only part running on GPU:

configuration | `CMSSW_15_0_0` | + #47226
---|---|---
full HLT menu | 661.2 ± 1.8 ev/s | 660.2 ± 0.8 ev/s
alpaka-only   | 567.7 ± 1.4 ev/s | 567.5 ± 0.9 ev/s

fwyzard commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1

makortel commented:
@fwyzard Should this PR be backported to 15_0_X?

fwyzard commented:
Yes, I'd sayit can be useful. 
I think you can reuse the same branch ?

makortel commented:
> Yes, I'd sayit can be useful.
> I think you can reuse the same branch ?

Yup, done in https://github.com/cms-sw/cmssw/pull/47515
